<a href="https://colab.research.google.com/github/asifahsaan/NLP-Natural-Language-Processing-/blob/main/WordSimilarity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center>**WORD SIMILARITY**</center>

<br>
<center>

|Title:     | Word Similarity|
|         :-            |                             :-:                               |
| **Resource Person:**  | Dr. Maryam                                      |
| **Course Title:**     | Natural Language Processing                                                  |
| **Credit Hours:**     | 3(3,0)                                                        |
| **Work Objective:**     | Improve Naive Method for Word Similarity                                                    |
<br>
            |          **Submitted By** |

|Sr. # | Student Name                   | Registration No. | Email Address           |
| :-   | :-                     | :-               | :-                      
| 1    | Asif Ahsan | 20L-2370         | l202370@lhr.nu.edu.pk |

</center>
<br>

### **Importing libraries**

In [ ]:
import numpy as np
import pandas as pd
from itertools import combinations
from difflib import SequenceMatcher
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

try:
  %load_ext google.colab.data_table
except:
  %reload_ext google.colab.data_table
  
import warnings
warnings.filterwarnings("ignore")

The google.colab.data_table extension is already loaded. To reload it, use:
  %reload_ext google.colab.data_table


## **Problem Statement**
The problem is to match the user's free-form input against a pre-determined list of banks. For example, user input 'bawag bank' should be matched to 'BAWAG Group AG'.

### Banks List for comparison

In [ ]:
# List of banks to compare
banks =   ['Sberbank Europe AG', 'BAWAG Group AG', 'Raiffeisenbankengruppe OÖ Verbund eGen', 'Raiffeisen Bank International AG', 'Volksbanken Verbund',
          'Erste Group Bank AG', 'KBC Groep', 'Investeringsmaatschappij Argenta', 'Belfius Bank', 'AXA Bank Belgium', 'The Bank of New York Mellon SA/NV',
          'First Investment Bank AD', 'RCB Bank Ltd', 'Bank of Cyprus Holdings Public Limited Company', 'Hellenic Bank Public Company Limited',
          'DekaBank Deutsche Girozentrale', 'Erwerbsgesellschaft der S-Finanzgruppe mbH & Co. KG', 'UBS Europe SE', 'DEUTSCHE APOTHEKER- UND ÄRZTEBANK EG',
          'Volkswagen Bank Gesellschaft mit beschränkter Haftung', 'Münchener Hypothekenbank eG', 'DZ BANK AG Deutsche Zentral-Genossenschaftsbank, Frankfurt am Main',
          'HASPA Finanzholding', 'State Street Europe Holdings Germany S.a.r.l. & Co. KG', 'J.P. Morgan AG', 'DEUTSCHE BANK AKTIENGESELLSCHAFT', 'COMMERZBANK Aktiengesellschaft',
          'Landesbank Baden-Württemberg', 'Landesbank Hessen-Thüringen Girozentrale', 'Norddeutsche Landesbank - Girozentrale -', 'Deutsche Pfandbriefbank AG',
          'Aareal Bank AG', 'Hamburg Commercial Bank AG', 'Bayerische Landesbank', 'Jyske Bank A/S', 'Sydbank A/S', 'Nykredit Realkredit A/S', 'Danske Bank A/S',
          'Luminor Holding AS',
          'Abanca Corporacion Bancaria S.A.',
          'Banco Santander S.A.',
          'Ibercaja Banco S.A.',
          'Kutxabank S.A',
          'Unicaja Banco S.A.',
          'CaixaBank S.A.',
          'Banco de Crédito Social Cooperativo',
          'Banco Bilbao Vizcaya Argentaria S.A.',
          'Banco de Sabadell S.A.',
          'Bankinter S.A.',
          'Kuntarahoitus Oyj',
          'Nordea Bank Abp',
          'OP Osuuskunta',
          'SFIL',
          'RCI Banque',
          'Confédération Nationale du Crédit Mutuel',
          'La Banque Postale',
          'Bpifrance',
          "C.R.H. - Caisse de refinancement de l'habitat",
          'HSBC Continental Europe',
          'Groupe BPCE',
          'Groupe Crédit Agricole',
          'Société générale',
          'BNP Paribas',
          'ALPHA SERVICES AND HOLDINGS S.A.',
          'National Bank of Greece S.A.',
          'Eurobank Ergasias Services and Holdings S.A.',
          'Piraeus Financial Holdings',
          'OTP-csoport',
          'Magyar Bankholding',
          'Barclays Bank Ireland plc',
          'Citibank Holdings Ireland Limited',
          'AIB Group plc',
          'Bank of Ireland Group plc',
          'Ulster Bank Ireland Designated Activity Company',
          'Bank of America Europe Designated Activity Company',
          'Íslandsbanki hf.',
          'Landsbankinn hf.',
          'Arion banki hf',
          'Intesa Sanpaolo S.p.A.',
          'Gruppo Bancario Finecobank  ',
          'UniCredit S.p.A.',
          'Gruppo Bancario Mediolanum  ',
          'Credito Emiliano Holding S.p.A.',
          'Banco BPM SpA',
          'Banca Popolare di Sondrio, Società Cooperativa per Azioni',
          'Banca Monte dei Paschi di Siena S.p.A.',
          'CASSA CENTRALE BANCA',
          'ICCREA BANCA S.P.A.',
          'Mediobanca - Banca di Credito Finanziario S.p.A.',
          'Akcine bendrove Šiauliu bankas',
          'Precision Capital S.A.',
          'RBC Investor Services Bank S.A.',
          'J.P. Morgan Bank Luxembourg S.A.',
          'Banque Internationale à Luxembourg',
          'Banque et Caisse d´Epargne de l´Etat, Luxembourg',
          'Akciju sabiedriba "Citadele banka"',
          'MDB Group Limited',
          'Bank of Valletta Plc',
          'HSBC Bank Malta p.l.c.',
          'BNG Bank N.V.',
          'ING Groep N.V.',
          'LP Group B.V.',
          'de Volksbank N.V.',
          'ABN AMRO Bank N.V.',
          'Coöperatieve Rabobank U.A.',
          'Nederlandse Waterschapsbank N.V.',
          'Bank Polska Kasa Opieki S.A.',
          'Powszechna Kasa Oszczednosci Bank Polski S.A.',
          'LSF Nani Investments S.à r.l.',
          'Banco Comercial Português SA',
          'Caixa Geral de Depósitos SA',
          'Banca Transilvania',
          'Länförsäkringar Bank AB (publ)',
          'Kommuninvest - group',
          'Skandinaviska Enskilda Banken - group',
          'SBAB Bank AB - group',
          'Swedbank - group',
          'Svenska Handelsbanken - group',
          'Biser Topco S.à r.l.',
          'Nova Ljubljanska Banka d.d. Ljubljana']

### **Old Code for Word Similarity**

In [ ]:
# Examples of search strings
s1 = 'Bawag bank' # other options: 'Bawag bank', 'Erste', 'Raiffaisen bank'

In [ ]:
# A naive search method which you need to improve
res = []
for token in banks:
  res.append([s1, token, SequenceMatcher(None, s1, token).ratio()])

df2 = pd.DataFrame(res, columns=['Bank 1', 'Bank 2', 'Score'])
# The outcome is not great, for this search query 'BAWAG Group AG' should have highest similarity
df2.sort_values(by=['Score'], ascending=[False]).head()

Bank 1                 Bank 2     Score
8   Bawag bank           Belfius Bank  0.454545
12  Bawag bank           RCB Bank Ltd  0.454545
33  Bawag bank  Bayerische Landesbank  0.451613
42  Bawag bank          Kutxabank S.A  0.434783
99  Bawag bank          BNG Bank N.V.  0.434783

In [ ]:
 #The desired combination has a low score
idx = df2['Bank 2'].isin(['BAWAG Group AG'])

df2[idx].sort_values(by=['Score'], ascending=[False]).head()

Bank 1          Bank 2     Score
1  Bawag bank  BAWAG Group AG  0.166667

### **Improved Word Similarity**
Now Testing on Given Data "Bawag Bank

In [ ]:
# Query
query = 'Bawag Bank'

# Create TF-IDF vectors for the banks
vectorizer = TfidfVectorizer()
bank_vectors = vectorizer.fit_transform(banks)

# Create a dataframe to store the results
similarity_df = pd.DataFrame(columns=['Bank 1', 'Bank 2', 'Score'])

# Calculate the cosine similarity between the query and each bank
query_vector = vectorizer.transform([query])
for i, bank in enumerate(banks):
    bank_vector = bank_vectors[i]
    similarity = cosine_similarity(query_vector, bank_vector)[0][0]
    similarity_df.loc[i] = [query, bank, similarity]

# Sort the results by similarity score
similarity_df = similarity_df.sort_values(by='Score', ascending=False)

# Print the results
similarity_df

Bank 1                                 Bank 2     Score
1    Bawag Bank                         BAWAG Group AG  0.663232
34   Bawag Bank                         Jyske Bank A/S  0.168777
8    Bawag Bank                           Belfius Bank  0.168777
99   Bawag Bank                          BNG Bank N.V.  0.168777
37   Bawag Bank                        Danske Bank A/S  0.168777
..          ...                                    ...       ...
46   Bawag Bank   Banco Bilbao Vizcaya Argentaria S.A.  0.000000
45   Bawag Bank    Banco de Crédito Social Cooperativo  0.000000
44   Bawag Bank                         CaixaBank S.A.  0.000000
43   Bawag Bank                     Unicaja Banco S.A.  0.000000
119  Bawag Bank  Nova Ljubljanska Banka d.d. Ljubljana  0.000000

[120 rows x 3 columns]

## **Results Conclusion**

In [ ]:
print("OLD SIMILARITY SCORE")
print(df2[idx].sort_values(by=['Score'], ascending=[False]).head())

print("\nIMPROVED SIMILARITY SCORE")
print(similarity_df[idx].sort_values(by=['Score'], ascending=[False]).head())

OLD SIMILARITY SCORE
       Bank 1          Bank 2     Score
1  Bawag bank  BAWAG Group AG  0.166667

IMPROVED SIMILARITY SCORE
       Bank 1          Bank 2     Score
1  Bawag Bank  BAWAG Group AG  0.663232


We can see that we have **improve the similarity score** for the given data **"Bawag Bank"** initially we had the output score of **0.166667** but we improve it and our similarity score reaches to **0.663232**. Which has the **highest similarity score** according to our query.